sw/host/opentitanlib/src/transport/hyperdebug/mod.rs

```diff
@@ -0,0 +1,320 @@
+    spis: RefCell<HashMap<u8, Rc<dyn Target>>>,
+    uarts: RefCell<HashMap<PathBuf, Rc<dyn Uart>>>,
+}
+
+impl Hyperdebug {
+    pub const VID_GOOGLE: u16 = 0x18d1;
+    pub const PID_HYPERDEBUG: u16 = 0x520e;
+
+    /// Establish connection with a particular HyperDebug.
+    pub fn open(usb_vid: Option<u16>, usb_pid: Option<u16>, usb_serial: &Option<String>)
+                -> Result<Self> {
+        let device = UsbBackend::new(
+            usb_vid.unwrap_or(Self::VID_GOOGLE),
+            usb_pid.unwrap_or(Self::PID_HYPERDEBUG),
+            usb_serial)?;
+
+        let path = PathBuf::from("/sys/bus/usb/devices");
+
+        let mut console_tty: Option<PathBuf> = None;
+        let mut spi_interface: Option<BulkInterface> = None;
+        let mut uart_ttys: HashMap<String, PathBuf> = HashMap::new();
+
+        let config_desc = device.active_config_descriptor()?;
+        // Iterate through each USB interface, discovering e.g. supported UARTs.
+        for interface in config_desc.interfaces() {
+            for interface_desc in interface.descriptors() {
+                if let Some(idx) = interface_desc.description_string_index() {
+                    if let Ok(interface_name) = device.read_string_descriptor_ascii(idx) {
+                        let ports = device.port_numbers()?.iter().map(
+                            |id| id.to_string()).collect::<Vec<String>>().join(".");
+                        let interface_path = path
+                            .join(format!("{}-{}", device.bus_number(), ports))
+                            .join(format!("{}-{}:{}.{}",device.bus_number(), ports,
+                                          config_desc.number(), interface.number()));
+                        // Check the ASCII name of this USB interface.
+                        if interface_name == "HyperDebug Shell" {
+                            // We found the "main" control interface of HyperDebug, allowing
+                            // textual commands to be sent, to e.g. manipoulate GPIOs.
+                            console_tty = Some(Self::find_tty(&interface_path)?)
+                        } else if interface_name.starts_with("UART") {
+                            // We found an UART forwarding USB interface.
+                            uart_ttys.insert(
+                                interface_name.clone(),
+                                Self::find_tty(&interface_path)?);
+                        } else if interface_name == "SPI" {
+                            // We found the SPI forwarding USB interface (this one interface
+                            // allows multiplexing physical SPI ports.)
+                            let mut in_endpoint: u8 = 0;
+                            let mut out_endpoint: u8 = 0;
+                            for endpoint_desc in interface_desc.endpoint_descriptors() {
+                                if endpoint_desc.transfer_type() == rusb::TransferType::Bulk {
```

Comment:
swap this around to

```rust
if transfer_type != Bulk { continue }
```

That said, this seems like it will overwrite `in_endpoint` and such multiple times. Is that intentional?

Reply:
Thanks, there is enough indentation already.  (I also removed two levels of indentation by using the trick of doing let x = match ... and then putting continue in the "unwanted" match clause.)

About the overwriting, I do rely on the USB interface having exactly one bulk endpoint in either direction, without attempting to detect if it has more than one of either.  After all, we are controlling the firmware on the USB device.